python/tank/authentication/shotgun_shared/saml2_sso.py

```diff
@@ -0,0 +1,576 @@
+    #
+    # QTimer callbacks
+    #
+    ############################################################################
+
+    def on_schedule_sso_session_renewal(self):
+        """
+        Called to trigger the session renewal.
+
+        The session renewal, via the off-screen QWebView, will be done at the
+        next time the application event loop does not have any pending events.
+        """
+        log.debug("==- on_schedule_sso_session_renewal")
+
+        self._sso_renew_timer.start()
+
+    def on_renew_sso_session(self):
+        """
+        Called to renew the current SSO session.
+
+        The renewal will be done via an off-screen QWebView. The intent is to
+        benefit from the saved session cookies to automatically trigger the
+        renewal without having the user having to enter any inputs.
+        """
+        log.debug("==- on_renew_sso_session")
+
+        self._sso_renew_watchdog_timer.start()
+
+        # We do not update the page cookies, assuming that they have already
+        # have been cleared/updated before.
+        self._view.page().mainFrame().load(self._session.host + URL_SAML_RENEW_PATH)
+
+    def on_renew_sso_session_timeout(self):
+        """
+        Called when the SSO session renewal is taking too long to complete.
+
+        The purpose of this callback is to stop the page loading.
+        """
+        log.debug("==- on_renew_sso_session_timeout")
+        # @FIXME: Not sure this is the proper thing to do
+        # self._view.page().triggerAction(QtWebKit.QWebPage.Stop)
+        self._dialog.reject()
+
+    ############################################################################
+    #
+    # Qt event handlers
+    #
+    ############################################################################
+
+    def on_load_finished(self, succeeded):
+        """
+        Called by Qt when the Web Page has finished loading.
+
+        The renewal process goes thru a number of redirects. We detect the
+        end of the process by checking the page loaded, as we know where we
+        expect to land in the end.
+
+        At that point, we stop the process by sending the 'accept' event to
+        the dialog. If the process is taking too long, we have a timer
+        (_sso_renew_watchdog_timer) which will trigger and attempt to cleanup
+        the process.
+        """
+        # log.debug("==- on_load_finished")
+
+        url = self._view.page().mainFrame().url().toString()
+        if (
+                self._session is not None and
+                url.startswith(self._session.host + URL_SAML_RENEW_LANDING_PATH)
+        ):
+            self.update_session_from_browser()
+            if self._session_renewal_active:
+                self.start_sso_renewal()
+
+            self._dialog.accept()
+
+        if not succeeded and url != "":
+            log.error("Loading of page \"%s\" generated an error." % url)
+            # @FIXME: Figure out proper way of handling error.
```

Comment:
Certainly.

The thing is, often there will be an error but it will be non-critical: a css file could not be loaded, or such. There is little we can do to get around missing files on the server side. It is the responsibility of the IdP/SSO provider to ensure that loading their page does not cause errors.

And for other errors, I have attempted to trap all that I could in the `on_http_response_finished` callback. If we do get a critical error here, it is not clear how to get back on our feet. Sometime even loading a blank page causes an error.

In my (albeit limited) experience, getting to that line has proven to be a non-critical situation.